File: lib/ast.ml
type unaryOp =
  | Not
  | Neg

type binaryOp =
  | Add
  | Mul
  | And
  | Or

type 'i expr =
  | Var of 'i
  | Int of int
  | Bool of bool
  | OpUnary of unaryOp * 'i expr
  | OpBinary of binaryOp * 'i expr * 'i expr
  | Closure of 'i * 'i expr
  | Application of 'i expr * 'i expr (* actually a binary operation, but make it a separate constructor since it has special behaviour *)
  | If of 'i expr * 'i expr * 'i expr

let string_of_expr (e : string expr) : string =
  let rec string_of_expr' (e : string expr) : string =
    match e with
    | Var x -> x
    | Int n -> string_of_int n
    | Bool b -> string_of_bool b
    | OpUnary (op, e) -> (match op with
        | Not -> "not " ^ string_of_expr' e
        | Neg -> "-" ^ string_of_expr' e)
    | OpBinary (op, e1, e2) -> (match op with
        | Add -> string_of_expr' e1 ^ " + " ^ string_of_expr' e2
        | Mul -> string_of_expr' e1 ^ " * " ^ string_of_expr' e2
        | And -> string_of_expr' e1 ^ " && " ^ string_of_expr' e2
        | Or -> string_of_expr' e1 ^ " || " ^ string_of_expr' e2)
    | Closure (x, e) -> x ^ " -> " ^ string_of_expr' e
    | Application (e1, e2) -> "(" ^ string_of_expr' e1 ^ ") (" ^ string_of_expr' e2 ^ ")"
    | If (e1, e2, e3) -> "if " ^ string_of_expr' e1 ^ " then " ^ string_of_expr' e2 ^ " else " ^ string_of_expr' e3
  in
  string_of_expr' e

File: src/ast.rs
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum UnaryOp {
    Not,
    Neg,
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinaryOp {
    Equal,
    Add,
    Mul,
    And,
    Or,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Expr {
    Var(String),
    Int(i32),
    Bool(bool),
    OpUnary(UnaryOp, Box<Expr>),
    OpBinary(BinaryOp, Box<Expr>, Box<Expr>),
    Closure(String, Box<Expr>),
    Application(Box<Expr>, Box<Expr>),
    Let(String, Box<Expr>, Box<Expr>),
    If(Box<Expr>, Box<Expr>, Box<Expr>),
}

impl fmt::Display for Expr {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Expr::Var(x) => write!(f, "{}", x),
            Expr::Int(n) => write!(f, "{}", n),
            Expr::Bool(b) => write!(f, "{}", b),
            Expr::OpUnary(op, e) => match op {
                UnaryOp::Not => write!(f, "not {}", e),
                UnaryOp::Neg => write!(f, "-{}", e),
            },
            Expr::OpBinary(op, e1, e2) => match op {
                BinaryOp::Equal => write!(f, "{} = {}", e1, e2),
                BinaryOp::Add => write!(f, "{} + {}", e1, e2),
                BinaryOp::Mul => write!(f, "{} * {}", e1, e2),
                BinaryOp::And => write!(f, "{} && {}", e1, e2),
                BinaryOp::Or => write!(f, "{} || {}", e1, e2),
            },
            Expr::Closure(x, e) => write!(f, "{} -> {}", x, e),
            Expr::Application(e1, e2) => write!(f, "({}) ({})", e1, e2),
            Expr::Let(x, e1, e2) => write!(f, "let {} = {} in {}", x, e1, e2),
            Expr::If(e1, e2, e3) => write!(f, "if {} then {} else {}", e1, e2, e3),
        }
    }
}
